File: uri/1029/fibonacci.py
MAX = 40

def fib(values, calls):
    values.append(0) # values[0]
    values.append(1) # values[1]

    calls.append(0) # calls[0]
    calls.append(0) # calls[1]
    calls.append(2) # calls[2]

    for i in range(2, MAX):
        values.append(values[i-1] + values[i-2])
    
    for i in range(3, MAX):
        calls.append(calls[i-1] + calls[i-2] + 2)

n = int(input())
cases = []

for i in range(0, n):
    cases.append(int(input()))

values = []
calls = []
fib(values, calls)

for i in range(0, n):
    print "fib({}) = {} calls = {}".format(cases[i], calls[cases[i]], \
                                           values[cases[i]])

File: uri/1029/fibonacci.c
#include <stdlib.h>
#include <stdio.h>

#define MAX 40

void fib(int *values, int *calls) {
    // initial values
    values[0] = 0;
    values[1] = 1;

    // initial # of calls
    calls[0] = 0;
    calls[1] = 0;
    calls[2] = 2;

    for(int i = 2; i < MAX; i++) {
        values[i] = values[i-1] + values[i-2];
        calls[i] = calls[i-2] + calls[i-1] + 2;
    }
}

void main() {
    int n_tests;
    scanf("%d", &n_tests);
    
    int cases[MAX];
    int calls[MAX];
    int values[MAX];

    for(int i = 0; i < n_tests; i++) {
        scanf("%d", &cases[i]);
    }

    fib(values, calls);

    for(int i = 0; i < n_tests; i++) {
        printf("fib(%d) = %d calls = %d\n", cases[i], calls[cases[i]], values[cases[i]]);
    }
}